src/async_fuse/memfs/s3_metadata.rs

```diff
@@ -543,152 +559,202 @@ impl<S: S3BackEnd + Sync + Send + 'static> MetaData for S3MetaData<S> {
+            match new_entry {
+                None => {
+                    // new_name does not exist under new_parent
+                    if exchange {
+                        // exchange is true, new name must exist
+                        return build_enoent(
+                            new_name,
+                            new_parent,
+                            new_parent_node.lock().await.get_name(),
+                        );
+                    }
+                    // exchange is false, so we can do rename directly
+                    if old_parent == new_parent {
+                        let mut raw_node = old_parent_node.lock().await;
+                        raw_node.get_dir_data_mut().remove(old_name);
+                        raw_node.insert_entry_for_rename(DirEntry::new(
+                            new_name.into(),
+                            Arc::clone(old_entry.file_attr_arc_ref()),
+                        ));
+                    } else {
+                        let mut old_raw_node = old_parent_node.lock().await;
+                        let mut new_raw_node = new_parent_node.lock().await;
+                        old_raw_node.get_dir_data_mut().remove(old_name);
+                        new_raw_node.insert_entry_for_rename(DirEntry::new(
+                            new_name.into(),
+                            Arc::clone(old_entry.file_attr_arc_ref()),
+                        ));
+                    }
+                }
+                Some(new_entry) => {
+                    // new_name exists under new_parent
+                    if exchange {
+                        // exchange is true, so we can do exchange
+                        {
+                            // Remove from old_parent
+                            let mut raw_node = old_parent_node.lock().await;
+                            raw_node.get_dir_data_mut().remove(old_name);
+                            raw_node.insert_entry_for_rename(DirEntry::new(
+                                new_name.into(),
+                                Arc::clone(old_entry.file_attr_arc_ref()),
+                            ));
+                        }
+                        {
+                            // Remove from new_parent
+                            let mut raw_node = new_parent_node.lock().await;
+                            raw_node.get_dir_data_mut().remove(new_name);
+                            raw_node.insert_entry_for_rename(DirEntry::new(
+                                old_name.into(),
+                                Arc::clone(new_entry.file_attr_arc_ref()),
+                            ));
+                        }
+                    } else {
+                        // exchange is false, replace
+                        {
+                            // Remove from old_parent
+                            let mut raw_node = old_parent_node.lock().await;
+                            raw_node.get_dir_data_mut().remove(old_name);
```

Comment:
Same as above, call `remove_entry_for_rename` here.

Reply:
Besides, if a file is replaced, it's unlinked from its parent. Therefore, you should handle related logic, such as deleting.

Reply:
Delete logical will be implement after storage impl.

Reply:
#456